lib/goals.js

```diff
@@ -251,6 +251,79 @@ class GoalFollow extends Goal {
   }
 }
 
+// Moves the bot to a position where it is able to interact with the target
+// block, even from a slight distance. Used to create a more natural approach
+// to interacting and mining without needing to be so close to the target.
+class GoalInteract extends MoveTo {
+  constructor (bot, x, y, z) {
+    super(x, y, z)
+    this.bot = bot
+  }
+
+  isEnd(node)
+  {
+      return this.rayTraceBlock(node);
+  }
+
+  rayTraceBlock(node, maxSteps = 256, vectorLength = 5 / 16) {
```

Comment:
You can use the RaycastIterator I added to mineflayer: https://github.com/PrismarineJS/mineflayer/blob/master/lib/iterators.js#L105
It computes the step so that each step advance exactly to the next block (so you don't need ridiculously small steps to not ghost through blocks), resulting in only 1 check per block on the line of sight.

Reply:
I can't use existing raycasters as the goal relies on what the state of the world *will* be at that point in time. The blocks to be broken haven't been mined yet, so the world raycaster will deny it even though it should pass.

Reply:
The iterator is not tied to a world, it only iterates on block positions. So you can use it in the context you want.

Reply:
Oh! How did I miss this? Wow, that's significantly more helpful, lol. I'll definitely be using this, then.

At first, I thought you were talking about this one: https://github.com/PrismarineJS/mineflayer/blob/master/lib/plugins/ray_trace.js

Reply:
yeah,  https://github.com/PrismarineJS/mineflayer/blob/master/lib/plugins/ray_trace.js should be replaced as well